File: evals/behavioral-analysis/data/go/template-injection/django_template_injection_custom_tags.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Django {% load %}-style marker + text/template SSTI.

package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	tmpltext "text/template"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type djangoRenderer struct{}

func (*djangoRenderer) renderTemplate(s string, ctx map[string]any, autoescape bool) (string, error) {
	_ = autoescape
	t, err := tmpltext.New("django").
		Funcs(tmpltext.FuncMap{
			"env": os.Getenv,
		}).
		Parse(s)
	if err != nil {
		return "", err
	}
	var buf bytes.Buffer
	if err := t.Execute(&buf, ctx); err != nil {
		return "", err
	}
	return buf.String(), nil
}

func (*djangoRenderer) renderWithCustomTags(s string, ctx map[string]any, tags string) (string, error) {
	body := s
	if tags != "" {
		body = "{% load " + tags + " %}\n" + s
	}
	return (&djangoRenderer{}).renderTemplate(body, ctx, true)
}

func (*djangoRenderer) renderFromString(s string, ctx map[string]any) (string, error) {
	return (&djangoRenderer{}).renderTemplate(s, ctx, true)
}

func main() {
	srv := server.NewMCPServer("template-injection-service-9", "1.0.0")

	tool := mcp.NewTool("render_django_template",
		mcp.WithDescription("Render Django template with context and custom tags."),
		mcp.WithString("template_string", mcp.Required()),
		mcp.WithString("context_json"),
		mcp.WithString("custom_tags"),
		mcp.WithBoolean("disable_autoescape"),
		mcp.WithBoolean("use_engine"),
	)

	srv.AddTool(tool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		tpl := req.GetString("template_string", "")
		dataJSON := req.GetString("context_json", "{}")
		tags := req.GetString("custom_tags", "")
		noEsc := req.GetBool("disable_autoescape", false)
		useEngine := req.GetBool("use_engine", false)

		var ctx map[string]any
		if err := json.Unmarshal([]byte(dataJSON), &ctx); err != nil {
			ctx = map[string]any{"content": dataJSON}
		}

		r := &djangoRenderer{}
		var (
			out string
			err error
		)

		switch {
		case tags != "":
			out, err = r.renderWithCustomTags(tpl, ctx, tags)
		case useEngine:
			out, err = r.renderFromString(tpl, ctx)
		default:
			out, err = r.renderTemplate(tpl, ctx, !noEsc)
		}

		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("Render error: %v", err)), nil
		}
		if len(out) > 500 {
			out = out[:500]
		}
		return mcp.NewToolResultText("Template rendered:\n" + out + "..."), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

File: evals/behavioral-analysis/data/rust/template-injection/django_template_injection_custom_tags.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Django-style {% load %} prefix + caller-controlled Tera body

use rmcp::{
    handler::server::tool::ToolRouter,
    tool, tool_handler, tool_router,
    ServerHandler,
    ServiceExt,
    transport::stdio,
};
use rmcp::model::{CallToolResult, Content};
use schemars::JsonSchema;
use serde::Deserialize;
use serde_json::Value;
use tera::{Context, Tera};

#[derive(Deserialize, JsonSchema)]
struct DjangoArgs {
    template_string: String,
    #[serde(default)] context_json: String,
    #[serde(default)] custom_tags: String,
    #[serde(default)] disable_autoescape: bool,
    #[serde(default)] use_engine: bool,
}

#[derive(Clone)]
struct DjangoSrv {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl DjangoSrv {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "Render Django template with context and custom tags.")]
    async fn render_django_template(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<DjangoArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let data: serde_json::Map<String, Value> =
            serde_json::from_str(&args.context_json).unwrap_or_else(|_| {
                let mut m = serde_json::Map::new();
                m.insert(
                    "content".into(),
                    Value::String(args.context_json.clone()),
                );
                m
            });

        let mut ctx = Context::new();
        for (k, v) in &data {
            ctx.insert(k, v);
        }

        let tpl = if !args.custom_tags.is_empty() {
            format!("{{% load {} %}}\n{}", args.custom_tags, args.template_string)
        } else {
            args.template_string.clone()
        };

        let auto_esc = !(args.disable_autoescape || args.use_engine);
        let rendered =
            Tera::one_off(&tpl, &ctx, auto_esc).unwrap_or_else(|e| format!("Render error: {e}"));

        let preview: String = rendered.chars().take(500).collect();
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Template rendered:\n{preview}..."
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for DjangoSrv {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = DjangoSrv::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
